R/legend.grob.R

```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
-							tck = if (is.null(legendi[['tck']])) { 0 } else { legendi[['tck']] },
-							at = do.breaks(c(0, legendi[['continuous.amount']]), legendi[['continuous.amount']]),
-							col = colorRamp,
-							width = if (!is.null(legendi[['angle']]) && legendi[['angle']] != 0) { legendi[['height']] } else { legendi[['width']] },
-							labels = list(
-								labels = if (is.null(legendi[['labels']])) { c('') } else { legendi[['labels']] },
-								at = if (is.null(legendi[['at']])) { NULL } else { legendi[['at']] },
-								cex = if (is.null(legendi[['cex']])) { 0.8 } else { legendi[['cex']] },
-								rot =  if (is.null(legendi[['labels.rot']])) { 0 } else { legendi[['labels.rot']] }
-								)
-							);
-
-
-						color.key.grob <- draw.colorkey(
-								key = legend.key,
-								draw = FALSE
-								);
-
-						# adjust justification to line up with key style legends
-						color.key.grob$framevp$layout$valid.just <- c(0,1);
-						color.key.grob$framevp$valid.just <- c(0,0.5);
-						
-						# need padding to line up with key style legends
-						color.key.grob$framevp$x = unit(0,'npc') + unit(1.69,'points');
-						
-						# set sizes
-						color.key.grob$framevp$height <- unit(legendi[['height']], 'lines');
-						color.key.grob$framevp$width <- unit(legendi[['width']], 'lines');
-
-						if(!is.null(legendi[['pos.x']])) {
-							color.key.grob$framevp$x <- unit(legendi[['pos.x']],'npc');
-							}
-
-						if(!is.null(legendi[['pos.y']])) {
-							color.key.grob$framevp$y <- unit(legendi[['pos.y']],'npc');
-							}
-
-						# Add the legend to the frame
-						legend.grob.final <- packGrob(
-							frame = legend.grob.final,
-							grob = color.key.grob,
-							row = 3 * (legend.row - 1) + 2,
-							col = 2 * (legend.col - 1) + 1,
-							height = max(unit(legendi[['height']], 'lines'),
-									legend.grob.final$framevp$layout$heights[3 * (legend.row - 1) + 1]),
-							force.height = TRUE,
-							);
-						}
-					else {
-						legend.key <- list(
-							just = c('left', 'top'),
-							between = 0.5,
-							rep = FALSE,
-							rectangles = list(
-								col = legendi[['colours']],
-								size = if (is.null(legendi[['size']])) { size } else { legendi[['size']] },
-								height = 1,
-								border = legendi[['border']]
-								),
-							text = list(
-								legendi[['labels']],
-								cex = label.cex,
-								fontfamily = font.family
-								)
-							);
-
-						# Add the legend to the frame
-						legend.grob.final <- packGrob(
-							frame = legend.grob.final,
-							grob = draw.key(key = legend.key, draw = FALSE),
-							row = 3 * (legend.row - 1) + 2,
-							col = 2 * (legend.col - 1) + 1
-							);
-						}
+			# Figure out where this legend should go in the layout (ordered row-wise)
+			legend.row <- if (i %% layout[1] == 0) { i / layout[1] } else { floor(i / layout[1]) + 1 }
+			legend.col <- if (i %% layout[1] == 0) { layout[1] } else { i %% layout[1] }
```

Comment:
This needs explanation, why are both rows layout[1] rather than one being layout[0] and layout[1] respectively?